### PR TITLE
K8sService class

### PR DESCRIPTION
**Pull request by alinabuzachis**
##### SUMMARY

This refactors the `perform_action()` logic from `common.py` into a separate K8sService class.

TODO:
- [x] Unit tests.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

service.py

### CODE REVIEW

File: plugins/module_utils/k8s/service.py [hunk 1]
```diff
@@ -0,0 +1,609 @@
+# Copyright: (c) 2021, Red Hat | Ansible
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
Comment (Akasurde) [inline plugins/module_utils/k8s/service.py:2]: all files in module_utils are licensed as BSD license unless it is carried over from existing work or GNU GPL v3+ license work. @alinabuzachis Could you please clarify this?

Comment (gravesm) [reply plugins/module_utils/k8s/service.py:2]: This is all based on code from common.py which is GPL.

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

File: plugins/module_utils/k8s/service.py [hunk 2]
```diff
@@ -0,0 +1,639 @@
...
+    # Handled in module setup
+    pass
+
+try:
+    from kubernetes.dynamic.resource import Resource, ResourceInstance
+except ImportError:
+    # These are defined only for the sake of Ansible's checked import requirement
+    Resource = Any  # type: ignore
+    ResourceInstance = Any  # type: ignore
+    pass
```
Comment (Akasurde) [inline plugins/module_utils/k8s/service.py:42]:
```suggestion
except ImportError:
    # These are defined only for the sake of Ansible's checked import requirement
    Resource = Any  # type: ignore
    ResourceInstance = Any  # type: ignore
```

**Review by Akasurde — APPROVED** [review]
rest LGTM

**Review by gravesm — COMMENTED** [review]
Really great work! Aside from the `find()` method, just a few small changes, mostly with type hinting.

File: plugins/module_utils/k8s/service.py [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+                )
+            except (ResourceNotFoundError, ResourceNotUniqueError):
+                pass
+        try:
+            return self.client.resources.get(
+                api_version=api_version, short_names=[kind]
+            )
+        except (ResourceNotFoundError, ResourceNotUniqueError):
+            if fail:
+                self.module.fail(
```
[the remark this answers is not in the document]
Comment (alinabuzachis) [reply plugins/module_utils/k8s/service.py]: I kept `if fail:` and raised `CoreException ` inside. Not sure it we want to keep `if fail:`.

File: plugins/module_utils/k8s/service.py [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+                        "Failed to gather information about %s(s) even"
+                        " after waiting for %s seconds" % (res.get("kind"), duration)
+                    )
+                satisfied_by.append(res)
+            return dict(resources=satisfied_by, api_found=True)
+        result = result.to_dict()
+
+        if "items" in result:
+            return dict(resources=result["items"], api_found=True)
+        return dict(resources=[result], api_found=True)
```
[the remark this answers is not in the document]
Comment (alinabuzachis) [reply plugins/module_utils/k8s/service.py]: @gravesm Thank you for reviewing.

Replaced the code with this one and added `field_selectors` as param for `wait()`, removed `definition` and added `name` and `namespace`. I adapted the waiter's unit tests.

I also fixed the Client class return statement.

Let me know if there is anything else I can improve.

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

File: plugins/module_utils/k8s/waiter.py [hunk 5]
```diff
@@ -128,17 +128,23 @@ def __init__(
 
     def wait(
         self,
-        definition: Dict,
+        name: str,
+        namespace: str,
```
Comment (gravesm) [inline plugins/module_utils/k8s/waiter.py:132]: Both `name` and `namespace` should be optional.

**Review by gravesm — COMMENTED** [review]
One small change and this should be ready to go.

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
```
TASK [Wait job's pod running] **************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: 'K8sAnsibleMixin' object has no attribute 'fail_json'
fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/home/zuul/.ansible/tmp/ansible-tmp-1642003075.027186-76924-193286977235598/AnsiballZ_k8s_info.py\", line 107, in <module>\n    _ansiballz_main()\n  File \"/home/zuul/.ansible/tmp/ansible-tmp-1642003075.027186-76924-193286977235598/AnsiballZ_k8s_info.py\", line 99, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/zuul/.ansible/tmp/ansible-tmp-1642003075.027186-76924-193286977235598/AnsiballZ_k8s_info.py\", line 47, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.kubernetes.core.plugins.modules.k8s_info', init_globals=dict(_module_fqn='ansible_collections.kubernetes.core.plugins.modules.k8s_info', _modlib_path=modlib_path),\n  File \"/usr/lib64/python3.8/runpy.py\", line 207, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.8/runpy.py\", line 97, in _run_module_code\n    _run_code(code, mod_globals, init_globals,\n  File \"/usr/lib64/python3.8/runpy.py\", line 87, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_k8s_info_payload_huvbkrhp/ansible_k8s_info_payload.zip/ansible_collections/kubernetes/core/plugins/modules/k8s_info.py\", line 205, in <module>\n  File \"/tmp/ansible_k8s_info_payload_huvbkrhp/ansible_k8s_info_payload.zip/ansible_collections/kubernetes/core/plugins/modules/k8s_info.py\", line 201, in main\n  File \"/tmp/ansible_k8s_info_payload_huvbkrhp/ansible_k8s_info_payload.zip/ansible_collections/kubernetes/core/plugins/modules/k8s_info.py\", line 161, in execute_module\n  File \"/tmp/ansible_k8s_info_payload_huvbkrhp/ansible_k8s_info_payload.zip/ansible_collections/kubernetes/core/plugins/module_utils/common.py\", line 396, in kubernetes_facts\n  File \"/tmp/ansible_k8s_info_payload_huvbkrhp/ansible_k8s_info_payload.zip/ansible_collections/kubernetes/core/plugins/module_utils/common.py\", line 461, in fail\nAttributeError: 'K8sAnsibleMixin' object has no attribute 'fail_json'\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```
cc @gravesm

**Comment by gravesm** [comment]
Yeah, this is fixed in main and should be irrelevant once we switch to the new refactored code. This problem specifically was one of the things the refactor was intended to address. A recheck should work as the underlying error is almost certainly a timeout/network issue.

**Comment by gravesm** [comment]
recheck

**Comment by alinabuzachis** [comment]
@gravesm can we gate this or is there anything else I can do?

**Review by gravesm — APPROVED** [review]
Yes! This one is ready!

**Pull request merged**